common/goos/Object.cpp

```diff
@@ -83,6 +83,31 @@ std::string object_type_to_string(ObjectType type) {
   }
 }
 
+// converts an object type to it's symbol name to be converted into a typespec
+// returns 'object' otherwise
+std::string object_type_to_type_name(ObjectType kind) {
+  switch (kind) {
+    case ObjectType::INTEGER:
+      return "uint";
+    case ObjectType::FLOAT:
+      return "float";
+    case ObjectType::CHAR:
+      return "uint";
+    case ObjectType::SYMBOL:
+      return "symbol";
+    case ObjectType::STRING:
+      return "string";
+    case ObjectType::PAIR:
+      return "pair";
```

Comment:
I think this isn't going to do the right thing for some constants. Eg:
```
(defglobalconstant foo (+ 1 2))
```
will be `pair` according to this function. I'd recommend just leaving off types for constants for now. GOAL doesn't really support typed constants.